src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/PackageRestoreBar.xaml.cs

```diff
@@ -145,7 +212,51 @@ private void UpdateRestoreBar(bool packagesMissing)
 
         private void OnRestoreLinkClick(object sender, RoutedEventArgs e)
         {
-            NuGetUIThreadHelper.JoinableTaskFactory.RunAsync(() => UIRestorePackagesAsync(CancellationToken.None)).PostOnFailure(nameof(PackageRestoreBar));
+            NuGetUIThreadHelper.JoinableTaskFactory.RunAsync(() =>
+            {
+                if (_projectContextInfo?.ProjectStyle == ProjectModel.ProjectStyle.PackageReference && _isAssetsFileMissing)
+                {
+                    return UIRestoreProjectAsync(CancellationToken.None);
+                }
+
+                return UIRestorePackagesAsync(CancellationToken.None);
+            }).PostOnFailure(nameof(PackageRestoreBar));
+        }
+
+        private async Task<bool> UIRestoreProjectAsync(CancellationToken token)
+        {
+            await NuGetUIThreadHelper.JoinableTaskFactory.SwitchToMainThreadAsync();
```

Comment:
It's probably best if we don't await the whole thing on the UI thread. Switch to the UI thread change the restore bar in a separate async method. 

According to the threading rules we should only switch when we need to https://github.com/microsoft/vs-threading/blob/main/doc/threading_rules.md.